src/manager.rs

```diff
@@ -1,640 +1,213 @@
     }
 }
 
 
-trait Property<T> {
-    fn property(&self, path: &str, interface: &str, name: &str) -> Result<T, String>;
+enum_from_primitive!{
+#[derive(Debug, PartialEq)]
+pub enum NetworkManagerState {
+    Unknown = 0,
+    Asleep = 10,
+    Disconnected = 20,
+    Disconnecting = 30,
+    Connecting = 40,
+    ConnectedLocal = 50,
+    ConnectedSite = 60,
+    ConnectedGlobal = 70,
 }
-
-
-impl<T> Property<T> for NetworkManager
-    where NetworkManager: VariantTo<T>
-{
-    fn property(&self, path: &str, interface: &str, name: &str) -> Result<T, String> {
-        let property_error = |details: &str| {
-            Err(format!("D-Bus get '{}'::'{}' property failed on '{}': {}",
-                        interface,
-                        name,
-                        path,
-                        details))
-        };
-
-        let path = self.with_path(path);
-
-        match path.get(interface, name) {
-            Ok(variant) => {
-                match NetworkManager::variant_to(variant) {
-                    Some(data) => Ok(data),
-                    None => property_error("wrong property type"),
-                }
-            }
-            Err(err) => {
-                match err.message() {
-                    Some(details) => property_error(details),
-                    None => property_error("no details"),
-                }
-            }
-        }
-    }
 }
 
-
-trait VariantTo<T> {
-    fn variant_to(value: Variant<Box<RefArg>>) -> Option<T>;
-}
-
-
-impl VariantTo<String> for NetworkManager {
-    fn variant_to(value: Variant<Box<RefArg>>) -> Option<String> {
-        variant_to_string(value)
+impl From<u32> for NetworkManagerState {
+    fn from(val: u32) -> NetworkManagerState {
+        NetworkManagerState::from_u32(val).expect("Invalid Network Manager State enum value")
     }
 }
 
-
-impl VariantTo<i64> for NetworkManager {
-    fn variant_to(value: Variant<Box<RefArg>>) -> Option<i64> {
-        variant_to_i64(value)
+impl From<NetworkManagerState> for u32 {
+    fn from(val: NetworkManagerState) -> u32 {
+        val as u32
     }
 }
 
 
-impl VariantTo<u32> for NetworkManager {
-    fn variant_to(value: Variant<Box<RefArg>>) -> Option<u32> {
-        variant_to_u32(value)
-    }
+enum_from_primitive!{
+#[derive(Debug, PartialEq)]
+pub enum Connectivity { // See https://bugzilla.gnome.org/show_bug.cgi?id=776848
+    Unknown = 0,
+    None = 1,
+    Portal = 2,
+    Limited = 3,
+    Full = 4,
 }
-
-
-impl VariantTo<bool> for NetworkManager {
-    fn variant_to(value: Variant<Box<RefArg>>) -> Option<bool> {
-        variant_to_bool(value)
-    }
 }
 
-
-impl VariantTo<Vec<String>> for NetworkManager {
-    fn variant_to(value: Variant<Box<RefArg>>) -> Option<Vec<String>> {
-        variant_to_string_vec(value)
+impl From<u32> for Connectivity {
+    fn from(val: u32) -> Connectivity {
+        Connectivity::from_u32(val).expect("Invalid Connectivity enum value")
```

Comment:
No panics in `From` impls (see "**Note: this trait must not fail.** If the conversion can fail, use TryFrom or a dedicated method which returns an Option<T> or a Result<T, E>." from `From` docs).